Add `@content` support to `tailwindcss`

RobinMalfait opened the pull request:
This PR adds `@content` support to `tailwindcss`'s core package. We will handle the `@content` and call the `onContentPath` function when it's encountered.

The `@tailwindcss/cli`, `@tailwindcss/vite` and `@tailwindcss/postcss` packages have to implement the `onContentPath` such that the necessary globs are scanned and watchers should be setup with this information.

Example usage:

```css
@content "../../packages/my-sibling-package/src/components/*.tsx";
```

If you are in a monorepo setup, then you could point to other packages if you want. Another common use case is for Laravel projects if you want to point to Laravel blade files since they won't be covered by Vite's module graph:

```css
/* ./resources/css/app.css */
@content "../views/*.blade.php"
```

Note: all globs are relative to the current file you are in.

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
add `@content` support in `tailwindcss`

Integrations have to implement the `onContentPath` to know about the
globs found in `@content`.

RobinMalfait committed:
remove `globs: true` option from `scanDir`

All integrations will make use of this, so let's get rid of the option
entirely.

RobinMalfait committed:
allow to pass in `globs`

On the Rust side, we will make sure to optimize the `globs` and merge it
with the generated globs from the auto content detection.

RobinMalfait committed:
add tests

RobinMalfait committed:
rename `auto_content.rs` to `scan_dir.rs`

This is not only about auto content detection anymore, but it's testing
the `scan_dir` functionality.

`scan_dir` contains auto content detection and will handle passed in
globs.

RobinMalfait committed:
add `glob-match`

RobinMalfait committed:
add `serial_test`

RobinMalfait committed:
use `to_string_lossy` instead of `path.display()`

adamwathan commented on crates/oxide/src/glob.rs:
```diff
@@ -60,7 +64,11 @@ fn get_fast_patterns(base_path: &Path, patterns: &Vec<String>) -> Vec<(PathBuf,
+                vec![PathBuf::from("/")]
+            } else {
+                vec![base_path.to_path_buf()]
+            };
```
Let's update this to assume we are always getting relative paths because handling absolute paths is a bit more complicated across different operating systems and I'd rather handle that as a separate scope of work 👍

RobinMalfait committed:
assume all globs are relative with forward slashes

This means that we don't need special casing of the absolute paths. And
more importantly, we don't need to worry about Windows in this case
either.

RobinMalfait committed:
fix typo + apply clippy suggestions

RobinMalfait committed:
return `globs` from `compile()`

RobinMalfait committed:
`scan_dir` globs accept `Vec<GlobEntry>` instead of just string globs

This now means that all the globs passed in are a `GlobEntry`, which
contains the `base` and the `glob`.

thecrypticace reviewed (approved):
The windows tests are failing but I think that's addressed by the other PR so this looks good afaict
